token/core/common/loaders.go

```diff
@@ -0,0 +1,223 @@
+		TokenVault:   tokenVault,
+		Deserializer: deserializer,
+		NumRetries:   3,
+		RetryDelay:   3 * time.Second,
+	}
+}
+
+// GetTokenOutputs takes an array of token identifiers (txID, index) and returns the corresponding token outputs
+func (s *VaultLedgerTokenLoader[T]) GetTokenOutputs(ids []*token.ID) ([]T, error) {
+
+	var err error
+	for i := 0; i < s.NumRetries; i++ {
+		tokens := make([]T, len(ids))
+		counter := 0
+		err = s.TokenVault.GetTokenOutputs(ids, func(id *token.ID, bytes []byte) error {
+			if len(bytes) == 0 {
+				return errors.Errorf("failed getting serialized token output for id [%v], nil value", id)
+			}
+			ti, err := s.Deserializer.DeserializeToken(bytes)
+			if err != nil {
+				return errors.Wrapf(err, "failed deserializing token for id [%v][%s]", id, string(bytes))
+			}
+			tokens[counter] = ti
+			counter++
+			return nil
+		})
+		if err == nil {
+			return tokens, nil
+		}
+
+		// check if there is any token id whose corresponding transaction is pending
+		// if there is, then wait a bit and retry to load the outputs
+		retry := false
+		for _, id := range ids {
+			pending, err := s.TokenVault.IsPending(id)
+			if err != nil {
+				break
+			}
+			if pending {
+				//logger.Warnf("failed getting serialized token output for id [%v] because the relative transaction is pending, retry at [%d]", id, i)
+				if i >= s.NumRetries-1 {
+					// too late, we tried already too many times
+					return nil, errors.Wrapf(err, "failed to get token outputs, tx [%s] is still pending", id.TxId)
+				}
+				time.Sleep(s.RetryDelay)
+				retry = true
+				break
+			}
+		}
+
+		if retry {
+			tokens = nil
+			continue
+		}
+
+		return nil, errors.Wrapf(err, "failed to get token outputs")
+	}
+
+	return nil, errors.Wrapf(err, "failed to get token outputs")
+}
+
+type VaultLedgerTokenAndMetadataLoader[T LedgerToken, M any] struct {
+	TokenVault   TokenVault
+	Deserializer TokenAndMetadataDeserializer[T, M]
+}
+
+func NewVaultLedgerTokenAndMetadataLoader[T LedgerToken, M any](tokenVault TokenVault, deserializer TokenAndMetadataDeserializer[T, M]) *VaultLedgerTokenAndMetadataLoader[T, M] {
+	return &VaultLedgerTokenAndMetadataLoader[T, M]{TokenVault: tokenVault, Deserializer: deserializer}
+}
+
+// LoadTokens takes an array of token identifiers (txID, index) and returns the keys in the vault
+// matching the token identifiers, the corresponding zkatdlog tokens, the information of the
+// tokens in clear text and the identities of their owners
+// LoadToken returns an error in case of failure
+func (s *VaultLedgerTokenAndMetadataLoader[T, M]) LoadTokens(ids []*token.ID) ([]string, []T, []M, []view.Identity, error) {
+	n := len(ids)
+	tokens := make([]T, n)
+	inputIDs := make([]string, n)
+	inputInf := make([]M, n)
+	signerIds := make([]view.Identity, n)
+
+	// return token outputs and the corresponding opening
+	counter := 0
+	if err := s.TokenVault.GetTokenInfoAndOutputs(ids, func(id *token.ID, key string, comm, info []byte) error {
```

Comment:
Maybe we should change the signature of this method to return directly the slices. Because now we get the slices, and we call a callback function on them. and then with this callback we put them back into slices.

Reply:
Same also for the other vault operations

Reply:
we need a deserialization step here, no?